drv/lpc55-spi-server/src/main.rs

```diff
@@ -8,24 +8,329 @@
+            rx,
+            count: 0,
+            end: 0, // Set to expected fixed length message or rx.len()
+        }
+    }
+
+    // Set Rx idle and ready to receive.
+    pub fn ready(&mut self) {
+        let mut msg = Msg::parse(&mut *self.rx).unwrap_lite();
+        msg.set_msgtype(drv_spi_msg::MsgType::Invalid);
+        let len = msg.payload_buf().len();
+        self.state = RxState::Idle;
+        ringbuf_entry!(Trace::ResetRxToIdle);
+        self.count = 0;
+        self.end = len;
+    }
+
+    /// Process an input byte.
+    /// A start of transmission condition resets the receive state machine.
+    pub fn rx_byte(&mut self, b: u8, sot: bool, tx: &mut TxContext) {
+        // SP can issue a command (or no-op) at any start of transmission event.
+        if sot {
+            ringbuf_entry!(Trace::RxState(self.state)); // log old state
+            self.ready();
+        }
+
+        match self.state {
+            RxState::Invalid => {
+                // Ignoring bytes due to previous error or completion of Rx.
+            }
+            RxState::Idle => {
+                // The first received byte must be a supported protocol number.
+                self.rx[self.count] = b;
+                self.count += 1;
+                match b {
+                    SPI_MSG_IGNORE => {
+                        // A zero byte marks the message as one to ignore.
+                        self.state = RxState::Invalid;
+                    }
+                    SPI_MSG_VERSION => {
+                        // The only supported protocol at this time is 0x01.
+                        // Collect a header.
+                        self.state = RxState::Header;
+                    }
+                    _ => {
+                        // Form an error response unless Tx is sending something
+                        // valid.
+                        if tx.enqueue(
+                            MsgType::ErrorRsp,
+                            EnqueueBuf::Copy(&self.rx[0..1]),
+                        ) {
+                            ringbuf_entry!(Trace::Enqueued);
+                            self.state = RxState::Dispatch; // or RxState::Error?
+                                                            // TODO: a documented error type should be the first byte.
+                        } else {
+                            // We received garbage at sot but Tx was not idle.
+                            ringbuf_entry!(Trace::EnqueueFail);
+                        }
+                    }
+                }
+            }
+            RxState::Header => {
+                self.rx[self.count] = b;
+                self.count += 1;
+                ringbuf_entry!(Trace::HeaderCollect);
+                if self.count == SPI_HEADER_SIZE {
+                    ringbuf_entry!(Trace::HeaderComplete);
+                    let max_len = self.rx.len();
+                    let msg = Msg::parse(&mut *self.rx).unwrap_lite();
```

Comment:
I don't love that `Msg::parse()` is literally always followed by `unwrap_lite` – that suggests that we could be encoding things differently.

Taking a step back, why does `RxContext` store a single slice of arbitrary size, rather than _owning_ separate slices for the header and contents, e.g.

```
struct RxContext<'a> {
    state: RxState,
    header: MsgHeader,
    buf: [u8; MAX_SPI_MSG_PAYLOAD_SIZE],
    count: usize, // cursor for receive
    end: usize,   // cursor for receive
}
```

When in the `Header` state, we could write directly to `self.header.as_bytes()[self.count]`; in `Payload` we write to `self.payload[self.count]`, and never have to do runtime checks of slice sizes (so the entire `struct Msg` is eliminated).

This is a _tiny_ bit tricker on the `Tx` side, but is still doable / worthwhile: we'd split `TxState::Responding` into `TxState::Header` and `TxState::Payload`, then write out either `self.header` or `self.payload` in the same way.

Reply:
Alternatively, you could make the `struct Msg` an owner of its data:
```
#[derive(FromBytes, AsBytes, Unaligned, Debug)]
#[repr(C)]
pub struct Msg {
    header: MsgHeader,
    payload: [u8; MAX_SPI_MSG_PAYLOAD_SIZE],
}
```
then use `msg.as_bytes()[...]` when reading / writing to it.  Either way, it eliminates the runtime checks in `Msg::parse`.

Reply:
I took your suggestions and am happier with the code now. Thanks.